astropy/table/soco.py

```diff
@@ -0,0 +1,173 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+The SCEngine class uses the `sortedcontainers` package to implement an
+Index engine for Tables.
+"""
+
+from collections import OrderedDict
+from sortedcontainers import SortedList
```

Comment:
This module needs to still be importable even if `sortedcontainers` is not installed.  Then if user tries to create an `SCEngine` instance it should raise ImportError (probably?) at that point with a message that `sortedcontainers` needs to be installed.  This is slightly different than the BST version, where there is always the built-in backup if no `bintrees` is available.

Standard idiom is to define a `HAS_SOCO` module global in a try import / except block.  You can then import that in testing to decide whether or not to run index tests with `SCEngine`.

Reply:
This seems strange to me. There can be no SCEngine without `sortedcontainers`. To cause the import error when the `SCEngine` is created, the import would have to be moved to its `__init__` method.

The BST version is also a bit odd to me because if you try to import `FastBST` without `bintrees` installed then it lies and gives `BST` instead. What's the use-case here: to allow code that uses `FastBST` to work without `bintrees`? How would those users become aware of the degraded performance? Maybe there should be a warning emitted?

Another idea, to implement similar functionality here, would be to move the `from sortedcontainers import SortedList` line to the bottom in a try/except such that if we catch an `ImportError` then we set `SCEngine = BST` or whatnot. Again, probably should emit a warning.

I'm happy to see it implemented whatever way. Just want to understand better.

Reply:
While I agree that it would be odd to have different APIs depending on the installed packages I also think that *silently* falling back to slower (and possibly differently behaving) implementations isn't great either.

So I'd say the module should always be importable and the SCEngine should always be present in the API we should actually error out or warn if it's used without having sortedcontainers installed.

Reply:
Ok, I defined HAS_SOCO and made the module always importable. If Sorted Containers is not installed then SCEngine is present but cannot be instantiated.